h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1758,90 +1764,121 @@ private void compactMoveChunks(Iterable<Chunk> move) {
             sync();
+
+            long originalFileSize = getFileLengthInUse();
             for (Iterator<Chunk> iter = move.iterator(); iter.hasNext(); ) {
                 Chunk chunk = iter.next();
-                if (fileStore.predictAllocation(chunk.len) >= leftmostBlock) {
+                // while chunks destination do not overlap with their
+                // source location as a group, we can move them directly
+                // to the target destination
+                if (fileStore.predictAllocation(chunk.len) + chunk.len > leftmostBlock) {
+                    // overlap encountered, now rest of the chunks
+                    // are going to be moved in two steps - first to EOF,
+                    // and only then to target destination
+                    // this is to avoid possible corruption on abrupt shutdown
                     break;
                 }
-                moveChunk(chunk, false);
+                moveChunk(chunk, 0);
                 iter.remove();
             }
 
+            boolean movedToEOF = false;
+            // we need to ensure that chunks moved within the following loop
+            // do not overlap with space just released by chunks moved above,
+            // if some of them happened to sit right at the end of used space,
+            // hence the need to reserve this area
             for (Chunk chunk : move) {
-                moveChunk(chunk, true);
+                moveChunk(chunk, originalFileSize);
+                movedToEOF = true;
             }
 
-            // update the metadata (store at the end of the file)
-            reuseSpace = false;
-            commit();
-            sync();
+            long biggestFileSize = getFileLengthInUse();
+            Chunk chunkWithMetaToMove = null;
+            if (movedToEOF) {
+                // update the metadata (store at the end of the file)
+                reuseSpace = false;
+                store(0, biggestFileSize);
+                sync();
 
-            Chunk chunk = lastChunk;
+                chunkWithMetaToMove = lastChunk;
+                biggestFileSize = getFileLengthInUse();
 
-            // now re-use the empty space
-            reuseSpace = true;
-            for (Chunk c : move) {
-                moveChunk(c, false);
+                // now re-use the empty space
+                reuseSpace = true;
+                for (Chunk c : move) {
+                    moveChunk(c, 0);
+                }
             }
 
             // update the metadata (within the file)
-            commit();
+            store(movedToEOF ? originalFileSize : leftmostBlock * BLOCK_SIZE, biggestFileSize);
             sync();
-            if (moveChunk(chunk, false)) {
-                commit();
+
+            // if all chunks where moved straight to their target locations, but that last chunk
+            // with updated metadata did not fit, now we need to move it again
+            assert lastChunk != null;
```

Comment:
Here we should release reserved area, then reserve area of the last chunk 1(created in line 1800) and the last chunk 2(created in line 1814), and try to move them.